languages/js/src/filter.test.ts

```diff
@@ -328,25 +263,150 @@ async function fixtures() {
     | OrgRole
     | Bar
     | Foo
-    | Num;
-
-  const execQuery = (q: SelectQueryBuilder<Resource>) => q.getMany();
-  const combineQuery = <T extends SelectQueryBuilder<Resource>>(a: T, b: T) => {
-    // this is kind of bad but typeorm doesn't give you a lot of tools
-    // for working with queries :(
-    const whereClause = (sql: string) => /WHERE (.*)$/.exec(sql)?.[1];
-    const bClause = whereClause(b.getQuery());
-    if (bClause) a = a.orWhere(bClause, b.getParameters());
-    const aClause = whereClause(a.getQuery());
-    if (aClause) a = a.where(`(${aClause})`, a.getParameters());
-    return a;
+    | Num
+    | Log;
+
+  const repositories: {[key: string]: Repository<Resource>} = {
+    'User': users,
+    'Repo': repos,
+    'Org': orgs,
+    'Issue': issues,
+    'RepoRole': repoRoles,
+    'OrgRole': orgRoles,
+    'Bar': bars,
+    'Foo': foos,
+    'Num': nums,
+    'Log': logs,
   };
 
-  // set global exec/combine query functions
-  oso.setDataFilteringQueryDefaults({ execQuery, combineQuery });
+  const classes: {[key: string]: any} = {
+    'User': User,
+    'Repo': Repo,
+    'Org': Org,
+    'Issue': Issue,
+    'RepoRole': RepoRole,
+    'OrgRole': OrgRole,
+    'Bar': Bar,
+    'Foo': Foo,
+    'Num': Num,
+    'Log': Log,
+  };
+
+  async function buildQuery(filter: Filter): Promise<SelectQueryBuilder<Resource>> {
+    var repo: Repository<Resource> = repositories[filter.model];
+    let query: SelectQueryBuilder<Resource> = repo.createQueryBuilder(filter.model);
+
+    // Sort relations so they are in a sql valid order
+    let relations = []
+    let seen = [filter.model]
+    while (filter.relations.length > 0) {
+      let rest = []
+      for (let rel of filter.relations) {
+        if (seen.includes(rel.fromTypeName)) {
+          seen.push(rel.toTypeName)
+          relations.push(rel)
+        } else {
+          rest.push(rel)
+        }
+      }
+      filter.relations = rest
+    }
+    filter.relations = relations
+
+    for (let rel of filter.relations) {
+      let type = filter.types[rel.fromTypeName]
+      let other = classes[rel.toTypeName];
+      let relation = (type[rel.fromFieldName] as SerializedRelation).Relation;
+      let join = `${rel.fromTypeName}.${relation.my_field} = ${rel.toTypeName}.${relation.other_field}`
+      query = query.innerJoin(other, rel.toTypeName, join)
+    }
+
+    let nextId = 0
+    function toSql(d: Datum, values: {[key: string]: any}): string {
+      if (d.hasOwnProperty("value")) {
+        let key = `${nextId}`
+        nextId += 1
+        values[key] = (d as Immediate).value
+        return `:${key}`;
+      } else {
+        let proj = d as Projection;
+        return `${proj.typeName}.${proj.fieldName}`
+      }
+    }
+
+    for (let cs of filter.conditions) {
+      let values = {}
+      let ands = ""
+      for (let c of cs) {
+        let op;
+        switch(c.cmp) {
```

Comment:
could replace the switch with a `const ops: {[key: string]: string} = { 'Eq': '=', ... };`?